Source/JavaScriptCore/wasm/WasmBBQJIT.cpp

```diff
@@ -5927,39 +5985,312 @@ class BBQJIT {
+        ControlData dataCatch(*this, BlockType::Catch, data.signature(), data.enclosedHeight());
+        dataCatch.setCatchKind(CatchKind::Catch);
+        if (ControlData::isTry(data)) {
+            ++m_callSiteIndex;
+            data.setTryInfo(data.tryStart(), m_callSiteIndex, data.tryCatchDepth());
+        }
+        dataCatch.setTryInfo(data.tryStart(), data.tryEnd(), data.tryCatchDepth());
+
+        dataCatch.addBranch(data.releaseJumps());
+        dataCatch.addBranch(m_jit.jump());
+        LOG_DEDENT();
+        LOG_INSTRUCTION("Catch");
+        LOG_INDENT();
+        emitCatchImpl(dataCatch, exceptionSignature, results);
+        data = WTFMove(dataCatch);
+        m_exceptionHandlers.append({ HandlerType::Catch, data.tryStart(), data.tryEnd(), 0, m_tryCatchDepth, exceptionIndex });
         return { };
     }
 
-    PartialResult WARN_UNUSED_RETURN addTry(BlockSignature, Stack&, ControlType&, Stack&) BBQ_STUB
-    PartialResult WARN_UNUSED_RETURN addCatch(unsigned, const TypeDefinition&, Stack&, ControlType&, ResultList&) BBQ_STUB
-    PartialResult WARN_UNUSED_RETURN addCatchToUnreachable(unsigned, const TypeDefinition&, ControlType&, ResultList&) BBQ_STUB
-    PartialResult WARN_UNUSED_RETURN addCatchAll(Stack&, ControlType&) BBQ_STUB
-    PartialResult WARN_UNUSED_RETURN addCatchAllToUnreachable(ControlType&) BBQ_STUB
-    PartialResult WARN_UNUSED_RETURN addDelegate(ControlType&, ControlType&) BBQ_STUB
-    PartialResult WARN_UNUSED_RETURN addDelegateToUnreachable(ControlType&, ControlType&) BBQ_STUB
-    PartialResult WARN_UNUSED_RETURN addThrow(unsigned, Vector<ExpressionType>&, Stack&) BBQ_STUB
-    PartialResult WARN_UNUSED_RETURN addRethrow(unsigned, ControlType&) BBQ_STUB
+    PartialResult WARN_UNUSED_RETURN addCatchToUnreachable(unsigned exceptionIndex, const TypeDefinition& exceptionSignature, ControlType& data, ResultList& results)
+    {
+        m_usesExceptions = true;
+        ControlData dataCatch(*this, BlockType::Catch, data.signature(), data.enclosedHeight());
+        dataCatch.setCatchKind(CatchKind::Catch);
+        if (ControlData::isTry(data)) {
+            ++m_callSiteIndex;
+            data.setTryInfo(data.tryStart(), m_callSiteIndex, data.tryCatchDepth());
+        }
+        dataCatch.setTryInfo(data.tryStart(), data.tryEnd(), data.tryCatchDepth());
+
+        dataCatch.addBranch(data.releaseJumps());
+        LOG_DEDENT();
+        LOG_INSTRUCTION("Catch");
+        LOG_INDENT();
+        emitCatchImpl(dataCatch, exceptionSignature, results);
+        data = WTFMove(dataCatch);
+        m_exceptionHandlers.append({ HandlerType::Catch, data.tryStart(), data.tryEnd(), 0, m_tryCatchDepth, exceptionIndex });
+        return { };
+    }
+
+    PartialResult WARN_UNUSED_RETURN addCatchAll(Stack& expressionStack, ControlType& data)
+    {
+        m_usesExceptions = true;
+        data.endBlock(*this, data, expressionStack, false, true);
```

Comment:
For now, I would like to keep it streamlined simple one. After OSR entry thing is done and once we see the final implementation, I would like to do some refactoring (until that, simple duplicate code is better for modifying some part of it, since anyway OSR entry code also need to touch some of exception code.).